ddtrace/tracer.py

```diff
@@ -634,59 +652,47 @@ def trace(self, name, service=None, resource=None, span_type=None):
             >>> with tracer.trace('web.request') as span:
                     # do something
 
-        Trace will store the current active span and subsequent child traces will
-        become its children::
+        Example of the automatic parenting::
 
-            parent = tracer.trace('parent')     # has no parent span
-            child  = tracer.trace('child')      # is a child of a parent
+            parent = tracer.trace('parent')     # has no parent span, becomes the active span
+            child  = tracer.trace('child')      # is a child of a parent, becomes the active span
             child.finish()
             parent.finish()
 
-            parent2 = tracer.trace('parent2')   # has no parent span
+            parent2 = tracer.trace('parent2')   # has no parent span as parent and child have finished
             parent2.finish()
         """
-
-        # retrieve the Context using the context provider and create
-        # a new Span that could be a root or a nested span
-        context = self.get_call_context()
         return self.start_span(
             name,
-            child_of=context,
+            child_of=self.active(),
             service=service,
             resource=resource,
             span_type=span_type,
+            activate=True,
         )
 
     def current_root_span(self):
         # type: () -> Optional[Span]
-        """Returns the root span of the current context.
+        """Returns the root span of the current execution.
 
         This is useful for attaching information related to the trace as a
         whole without needing to add to child spans.
 
-        Usage is simple, for example::
+        For example::
 
             # get the root span
             root_span = tracer.current_root_span()
             # set the host just once on the root span
             if root_span:
                 root_span.set_tag('host', '127.0.0.1')
         """
-        ctx = self.get_call_context()
-        if ctx:
-            return ctx.get_current_root_span()
-        return None
+        parent = self.current_span()
+        if not parent:
+            return None
 
-    def current_span(self):
-        # type: () -> Optional[Span]
-        """
-        Return the active span for the current call context or ``None``
-        if no spans are available.
-        """
-        ctx = self.get_call_context()
-        if ctx:
-            return ctx.get_current_span()
-        return None
+        while parent._parent:
+            parent = parent._parent
+        return parent
```

Comment:
Alternative approaches:

- Each span has a local_root_span property which is passed downwards
- ContextVar also maintains a local root span reference

Reply:
I know once I wrote something like that and I got bitten because a user had 20k spans-long traces… I don't know how it applies here, but it seems to be pretty common to have worst-case scenarios in the trace data structure form. 😢

Reply:
Yeah it's probably worth the memory overhead to save on cpu here. The only argument against it is that accessing the "root" span is probably something we want to discourage users from doing and so this should be a seldomly used method (we do use it in a couple of integrations unfortunately).

Reply:
agreed, it isn't a great idea, but we don't have better ways to associate "trace metadata" other than just sticking it on the root span right now. if we can solve for those API usages in a more optimal way 👍🏻 (not in this PR obviously)

Reply:
This is resolved in 004e74d which follows the first approach "Each span has a local_root_span property which is passed downwards".